added blog post about workflow as a function flow on top of Knative E… (#4541)
knative/docs: blog/docs/articles/workflow-as-function-flow.md

	spec:
	  template:
	    metadata:
	      annotations:
	        autoscaling.knative.dev/minScale: "1"
	    spec:
	      containers:
	        - image: quay.io/ruben/cloudevents-player:latest
	          env:
	            - name: PLAYER_MODE
	              value: KNATIVE
	            - name: PLAYER_BROKER
	              value: default
	---
	apiVersion: eventing.knative.dev/v1
	kind: Trigger
	metadata:
	  name: cloudevents-player
	  annotations:
	    knative-eventing-injection: enabled
	spec:
	  broker: default
	  subscriber:
	    ref:
	      apiVersion: serving.knative.dev/v1
	      kind: Service
	      name: cloudevents-player
	EOF
	```

3. Get the url of the default broker use following command

	```
	kubectl get broker default
	```

4. Send request to the broker to start user registration

	```
	curl -v "http://broker-ingress.knative-eventing.svc.cluster.local/knative-eventing/default" \
	-X POST \
	-H "Ce-Id: 1234" \
	-H "Ce-Specversion: 1.0" \
	-H "Ce-Type: io.automatiko.examples.userRegistration" \
	-H "Ce-Source: curl" \
	-H "Content-Type: application/json" \
	-d '{"user" : {"email" : "mike.strong@email.com",  "firstName" : "mike",  "lastName" : "strong"}}'
	```

> 	This will go via number of events being exchanged over the Knative broker and invoking functions defined in the workflow. It also uses Swagger Petstore REST api so in case of successful user registration it will be visible in Swagger Petstore as new user.
> 
> 	**NOTE** That Swagger Petstore does not have reliable storage thus it might require few get requests to be issued to see it there.

Reviewer: ```suggestion
	This will go by number of events being exchanged over the Knative broker and invoking functions defined in the workflow. It also uses Swagger Petstore REST API, so in case of successful user registration it will be visible in Swagger Petstore as new user.

	!!! note
	    That Swagger Petstore does not have reliable storage thus it might require few get requests to be issued to see it there.
```

Reviewer: please check out https://knative.dev/docs/help/contributor/mkdocs-contributor-guide/#admonitions for info about notes etc

#### Clean up

To clean up, execute following command

````
kubectl delete -f target/functions/user-registration-1.0.0.yaml
````

### Build the service as container image for Google Cloud Run

To be able to use the same service for Google Cloud Run there are two additional steps required

1. Add dependency to `pom.xml`

	````
	<dependency>
	  <groupId>io.automatiko.extras</groupId>
	  <artifactId>automatiko-gcp-pubsub-sink</artifactId>
	</dependency>
	````

2. Add two properties to `src/main/resources/application.properties` file

	````
	quarkus.automatiko.target-deployment=gcp-pubsub
	quarkus.google.cloud.project-id=CHANGE_ME
	````

> 	**NOTE**: Remember to change `CHANGE_ME` to actual project id of your Google Cloud project

Reviewer: same formatting here for note

3. Build the application with following maven command

	````
	mvnw clean package -Pcontainer-native
	````

> 	This will build a container with the service that is using native executable built with GraalVM so the build process will take a while but it will be really light and fast at execution.

Reviewer: Same as earlier text, I'd edit this the same way and leave out the last bit, just make users aware it might take a while

Author: changed

gcloud eventarc triggers create generateusernameandpassword --event-filters="type=google.cloud.pubsub.topic.v1.messagePublished" --destination-run-service=user-registration-gcp-cloudrun --destination-run-path=/ --transport-topic=io.automatiko.examples.userRegistration.generateusernameandpassword --location=us-central1
gcloud eventarc triggers create getuser --event-filters="type=google.cloud.pubsub.topic.v1.messagePublished" --destination-run-service=user-registration-gcp-cloudrun --destination-run-path=/ --transport-topic=io.automatiko.examples.userRegistration.getuser --location=us-central1
gcloud eventarc triggers create userregistration --event-filters="type=google.cloud.pubsub.topic.v1.messagePublished" --destination-run-service=user-registration-gcp-cloudrun --destination-run-path=/ --transport-topic=io.automatiko.examples.userRegistration --location=us-central1

gcloud run deploy user-registration-gcp-cloudrun --platform=managed --image=gcr.io/CHANGE_ME/user/user-registration-gcp-cloudrun:1.0.0 --region=us-central1
````

> **NOTE**: `CHANGE_ME` will be replaced with the Google Cloud project id configured in `src/main/resources/application.properties` during the build.

Reviewer: note formatting again

> This will provision all required components such as

Reviewer: ```suggestion
This provisions all required components such as:
```

- PubSub Topics
- Eventarc Triggers
- Service deployment

#### Run the service on Google Cloud Run

> Once service is deployed you can publish first message (for example using Google Console) to the `io.automatiko.examples.userRegistration` topic with following content.

Reviewer: ```suggestion
After the service is deployed you can publish first message. For example, by using Google Console with the following `io.automatiko.examples.userRegistration` topic:
```

```
{"user" : {"email" : "mike.strong@email.com",  "firstName" : "mike",  "lastName" : "strong"}}
```

> 
> This will go via number of events being exchanged over the Google Cloud PubSub topics and invoking functions defined in the workflow.
> 
> 

Reviewer: ```suggestion

This goes by the number of events being exchanged over the Google Cloud PubSub topics and invoking functions defined in the workflow.

```

## Summing up

> Workflow as a function flow implements common scenario for serverless usage where
> individual functions build up a complete business case. It allows to use workflow to design complete business logic and then slice it into functions that are composed into a function flow based on the actual logic defined by the workflow definition. Yet each function can be invoked at anytime making the workflow to act as a blueprint that can start at any place and continue according to defined flow of activities aka functions.

Reviewer: ```suggestion
Workflow as a function flow implements common scenario for serverless usage where individual functions build up a complete business case. It allows you to use a workflow to design complete business logic and then slice it into functions that are composed into a function flow, based on the actual logic defined by the workflow definition. Yet each function can be invoked at anytime making the workflow to act as a blueprint that can start at any place and continue according to defined flow of functions.
```

> Workflow as a function flow takes advantage of Knative Eventing as the backbone of the communication to enable
> 
> - scalability as each function is invoked via Knative broker
> - all data exchange is done with Cloud Events
> - flexibility at which point in the workflow definition instance should be started
> - use "listen to yourself" principle to avoid long running actions impacting overall performance

Reviewer: ```suggestion
Workflow as a function flow takes advantage of Knative Eventing as the backbone of the communication to enable:

- Scalability as each function is invoked by using the Knative broker.
- All data exchange is done with Cloud Events.
- Flexibility with regard to which point in the workflow definition instance should be started.
- Use "listen to yourself" principle to avoid long running actions impacting overall performance.
```

---
title: Workflow as a Function Flow with Automatiko
linkTitle: Knative Eventing backed Workflow as a Function Flow
author: "[Maciej Swiderski](https://twitter.com/SwiderskiMaciek), Software Engineer @ OpenEnterprise."
date: 2021-08-20
description: Slicing business logic into functions and events.
type: "blog"

> ---

Reviewer: @snneji is this no longer required for blogs? It looks like hugo frontmatter?

Reviewer 2: Sorry for the late reply. I don't think this is required but I think it's useful information. In my opinion, the blog should show some of this information, however, it currently doesn't. I think we should keep it in case in the future we are able to show info like date, author, etc.

Various organisations started to look into serverless as a way of building business logic that can take advantage of the cloud. As it might look at first, it's not an easy task to rely strictly on functions that represent independent logic pieces. There is a risk of losing the big picture and by that not having full control over day-to-day operations.

> Knative comes with a great foundation to built upon when thinking about serverless, functions
> and events. Especially the Knative Eventing shines thanks to its integration with Cloud Events
> and various broker implementations that provide different characteristics from event delivery standpoint.

Reviewer: ```suggestion
Knative provides a great foundation to build upon when thinking about serverless, functions
and events. Knative Eventing especially shines thanks to its integration with Cloud Events
and various broker implementations that provide different characteristics from event delivery standpoint.
```

> At the same time, workflows (in various formats) have been seen as a good way to express business logic that allows to understand the big picture. Combining both (workflows and Knative Eventing) is the main topic of this article.
> 
> 
> ## Worfklow as a Function Flow

Reviewer: ```suggestion
At the same time, workflows in various formats have been seen as a good way to express business logic that enables understanding of the big picture. Combining both workflows and Knative Eventing is the main topic of this article.

## Worfklow as a function flow
```

> Workflows have been around for quite some time. It dates from the SOA (Service Oriented Architecture) times where standards such as BPEL and then BPMN were introduced. Various vendors started to build platforms that were built around the workflow concepts many times called Process Management Suites or Business Process Management (in short BPM).

Reviewer: ```suggestion
Workflows date from the Service Oriented Architecture (SOA) times, when standards such as BPEL and then BPMN were introduced. Various vendors started to build platforms that were built around the workflow concepts, often called Process Management Suites or Business Process Management (BPM).
```

> Most of the cases, these platforms are big, centralised servers that aim at orchestrating all other systems in a standardised way. That turned out to be too complex to bring value and in many cases is referred as traditional BPM which is already a thing of the past (even though still in use).

Reviewer: ```suggestion
These platforms are often big, centralised servers that aim to orchestrate all other systems in a standardised way. This turned out to be too complex to bring value, and in many cases is referred as traditional BPM, which is already outdated, although these are still in use today.
```

> Newer approaches are moving workflows to a different level, for example to make workflows completely distributed to the extend that each activity of the workflow runs as separate container.

Reviewer: ```suggestion
Newer approaches make workflows completely distributed, so that each activity of the workflow runs as separate container.
```

> This bring us to the concept of Workflow as a Function Flow. The main idea behind it is to allow users to model the business logic as complete as possible - meaning to define the business use case from the beginning till the end but take advantage of capabilities of the serverless platforms like Knative. That means that at runtime the workflow logic will be executed as set of functions that are

Reviewer: ```suggestion
This bring us to the concept of workflow as a function flow, which allows users to model the business logic as completely as possible. Users can define the business use case from the beginning until the end, and take advantage of capabilities of serverless platforms like Knative. That means that at runtime the workflow logic will be executed as a set of functions that are:
```

- **independent** - they rely on data given to execute logic defined and return outputs regardless if they were dealing with same instance or a different one
- **invokable at any time** - can be invoked at any time meaning that they do not have to always start from the beginning of the workflow as they simply react to incoming events based on the correlation attributes (Cloud Event type attribute)
- **scalable** - as functions can be scaled up and down easily to adopt to demands

![Workflow as Function Flow](/blog/articles/images/workflow-function-flow-concept.png)
*Workflow as Function Flow*

[Automatiko](https://automatiko.io) project delivers implementation of this concepts that takes advantage of Knative Eventing, Cloud events and workflow definitions to provide way of building your service as complete business use case but execute it as function flow that is controlled by the workflow definition but invoked by published events.

Automatiko supports following types of workflows

- [BPMN2 (Business Process Modeling and Notation)](https://www.bpmn.org)
- [Serverless Workflow Specification](https://serverlessworkflow.io)

### Slicing workflow into functions

> Functions are built based on executing activities - activities that can change workflow state and data. Such activities will then represent a particular fragment of business logic that will be encapsulated as function. Executing activities can

Reviewer: ```suggestion
Functions are built based on executing activities that can change workflow state and data. Such activities will then represent a particular fragment of business logic that will be encapsulated as function.

Executing activities can:
```

- be single activity
- include other activities to control logic

> - be combined with other (executing) activities to form so called continuation

Reviewer: ```suggestion
- be combined with other executing activities to form continuation
```

Reviewer: leave out phrases like "so called" in tech blogs ;)

Functions are named automatically based on workflow or activity names or explicitly by user using custom attributes.

> Each function will then become a dedicated entry point that will
> 
> - have input that is built based on workflow data
> - can produce one or more outputs
> - will have Knative trigger associated with the function

Reviewer: ```suggestion
Each function then becomes a dedicated entry point, which:

- has input that is built based on workflow data
- can produce one or more outputs
- has a Knative trigger associated with the function
```

![Slice workflow into functions](/blog/articles/images/workflow-function-flow-function-slice.png)
*Slice workflow into functions*

### Inputs and outputs as Cloud Events

> Once the functions are derived from the workflow, they will always have one event as input
> and can produce one or more events as outputs. The logic defined in the workflow and actual data context of a particular workflow instance will drive the number of events produced. Regardless of the event being an input or output it is

Reviewer: ```suggestion
After the functions are derived from the workflow, they will always have one event as input, and can produce one or more events as outputs. The logic defined in the workflow and actual data context of a particular workflow instance drives the number of events produced.

Regardless of whether the event is an input or output:
```

> - built based on data objects defined in the workflow
> - filed automatically based on execution context
> - uses **type** attribute as link to function
> - uses **source** attribute as identifier of the function being executed including unique identifier of workflow instance for correlation purposes

Reviewer: ```suggestion
- it is built based on data objects defined in the workflow
- it is filed automatically based on execution context
- it uses **type** attribute as link to function
- it uses **source** attribute as identifier of the function being executed including unique identifier of workflow instance for correlation purposes
```

![Inputs and outputs as events](/blog/articles/images/workflow-function-flow-function-io.png)
*Inputs and outputs as events*

> Functions always exchange events through Knative Eventing broker. They never call each other directly to make sure they are completely decoupled and event driven. This will allow for greater scalability as each function call can be handled by another replica of the service.

Reviewer: ```suggestion
Functions always exchange events through the Knative Eventing broker. They never call each other directly, which ensures that they are completely decoupled and event driven. This allows for greater scalability, as each function call can be handled by another replica of the service.
```

### Deployment

> First of all, the heavy work that requires parsing workflow definition, slicing that into
> functions and events and packaging into deployable unit is done at build time.
> 
> Automatiko project that implements this concept is a java based implementation that takes advantage of [GraalVM](https://www.graalvm.org) to be able to compile into native executable that brings small memory footprint at runtime and lightning fast startup time.
> 
> With that said, workflow as a function flow with Automatiko compiles into single service and by that into a single container image, that includes all the functions. Each and every function can be invoked at any time on any replica of the service.
> 
> At the same time (during build) Knative manifest file is generated to ease deployment to the cluster. That consists of
> 
> - Knative serving service
> - sink binding to inject broker location
> - triggers for each function created out of the workflow
> 
> So this serves as a starting point to have fully runnable service directly after build. It can be modified further to adopt to particular configuration of Knative in the cluster.
> 
> Enough theory, let's look at example of this concept.

Reviewer: ```suggestion
The heavy work that requires parsing workflow definition, slicing that into functions and events, and packaging this in to a deployable unit, is done at build time.

The Automatiko project implements this concept. It is a java based implementation that takes advantage of [GraalVM](https://www.graalvm.org) to compile into a native executable, which has a small memory footprint at runtime and lightning fast startup time.

Workflow as a function flow with Automatiko compiles into single service and by that into a single container image, that includes all the functions. Each and every function can be invoked at any time on any replica of the service.

At the same time during build, a Knative manifest file is generated to ease deployment to the cluster. The Knative manifest file consists of:

- a Knative serving service
- a sink binding to inject broker location
- triggers for each function created out of the workflow

This serves as a starting point to have fully runnable service directly after build. It can be modified further to adopt to particular configuration of Knative in the cluster.
```

## User registration example

This example it is a simple user registration that performs various checks and registers user in the Swagger PetStore service.

> See complete description of this example [here](https://docs.automatiko.io/main/0.0.0/examples/userregistration.html)

Reviewer: ```suggestion
You can see more information about this example in the [Automatiko documentation](https://docs.automatiko.io/main/0.0.0/examples/userregistration.html).
```

This example comes with two flavours (depending of what DSL is used to create workflow definition)

![Slice workflow into functions](/blog/articles/images/workflow-function-flow-function-example-bpmn.png)
*BPMN*

![Slice workflow into functions](/blog/articles/images/workflow-function-flow-function-example-sw.png)
*Serverless Workflow Spec*

> To try out the example yourself clone one of the flavours of this example project

Reviewer: ```suggestion
To try out the example yourself, clone one of the flavours of this example project:
```

* [User Registration BPMN](https://github.com/automatiko-io/automatiko-examples/tree/main/user-registration)
* [User Registration Serverless Workflow](https://github.com/automatiko-io/automatiko-examples/tree/main/user-registration-sw)
* [User Registration BPMN on Google Cloud Run](https://github.com/automatiko-io/automatiko-examples/tree/main/user-registration-gcp-cloudrun)

and build the application by issuing following maven command from within the cloned repository

````
mvn clean package -Pcontainer-native
````

This will build a container with the service that is using native executable built with GraalVM so the build process will take a while but it will be really light and fast at execution.

### Push built container to registry

> Once the container image is built push it to external registry that your Knative cluster can pull from.

Reviewer: ```suggestion
After the container image is built, push it to an external registry that your Knative cluster can pull from.
```

### Deploy to Knative cluster

> Create a Knative eventing broker, for example by using following command:

Reviewer: These should be numbered steps for a procedure, FYI

Author: changed to numbered steps
